pinot-core/src/main/java/com/linkedin/pinot/core/plan/maker/PlanNodeFactory.java

```diff
@@ -0,0 +1,161 @@
+          return new MetadataBasedAggregationPlanNode(indexSegment, brokerRequest.getAggregationsInfo());
+        } else if (isFitForDictionaryBasedPlan(brokerRequest, indexSegment)) {
+          return new DictionaryBasedAggregationPlanNode(indexSegment, brokerRequest.getAggregationsInfo());
+        } else {
+          return new AggregationPlanNode(indexSegment, brokerRequest);
+        }
+      }
+    }
+    if (brokerRequest.isSetSelections()) {
+      return new SelectionPlanNode(indexSegment, brokerRequest);
+    }
+    throw new UnsupportedOperationException("The query contains no aggregation or selection.");
+  }
+
+  /**
+   * Helper method to identify if query is fit to be be served purely based on metadata.
+   * Currently count queries without any filters, and min max queries if the segment has min max in
+   * metadata, are supported.
+   * @param brokerRequest Broker request
+   * @param indexSegment
+   * @return True if query can be served using metadata, false otherwise.
+   */
+  public static boolean isFitForMetadataBasedPlan(BrokerRequest brokerRequest, IndexSegment indexSegment) {
+    if (brokerRequest.getFilterQuery() != null || brokerRequest.isSetGroupBy()) {
+      return false;
+    }
+
+    List<AggregationInfo> aggregationsInfo = brokerRequest.getAggregationsInfo();
+    if (aggregationsInfo == null) {
+      return false;
+    }
+    for (AggregationInfo aggInfo : aggregationsInfo) {
+      if (!isMetadataBasedAggregationFunction(aggInfo, indexSegment)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static boolean isMetadataBasedAggregationFunction(AggregationInfo aggregationInfo,
+      IndexSegment indexSegment) {
+    String aggregationType = aggregationInfo.getAggregationType();
+    if (aggregationType.equalsIgnoreCase(AggregationFunctionFactory.AggregationFunctionType.COUNT.getName())) {
+      return true;
+    }
+
+    // Use minValue and maxValue from metadata, in non star tree cases
+    // minValue and maxValue is generated from dictionary on segment load, so it won't be correct in case of star tree
+    if ((aggregationType.equalsIgnoreCase(AggregationFunctionFactory.AggregationFunctionType.MAX.getName())
+        || aggregationType.equalsIgnoreCase(AggregationFunctionFactory.AggregationFunctionType.MIN.getName())
+        || aggregationType.equalsIgnoreCase(AggregationFunctionFactory.AggregationFunctionType.MINMAXRANGE.getName()))
+        && !indexSegment.getSegmentMetadata().hasStarTree()) {
+      String column = aggregationInfo.getAggregationParams().get("column");
+      SegmentMetadataImpl segmentMetadata = (SegmentMetadataImpl) indexSegment.getSegmentMetadata();
+      if (segmentMetadata.getColumnMetadataMap() == null) {
+        return false;
+      }
+      ColumnMetadata columnMetadata = segmentMetadata.getColumnMetadataFor(column);
+      if (columnMetadata.getMaxValue() != null && columnMetadata.getMinValue() != null) {
```

Comment:
what happens if min & max are not numeric. Will maxValue/minValue returned be null. Did we test that this works fine with old segments where we don't have min and max in the metadata.

Reply:
Works fine with older segments yes. Disabled  the path to min max from metadata for now.